graphix/opengraph.py

```diff
@@ -116,3 +118,75 @@ def to_pattern(self) -> Pattern:
         planes = {node: m.plane for node, m in meas.items()}
 
         return generate_from_graph(g, angles, inputs, outputs, planes)
+
+    def compose(self, other: OpenGraph, mapping: Mapping[int, int]) -> tuple[OpenGraph, dict[int, int]]:
+        r"""Compose two open graphs by merging subsets of nodes from `self` and `other`, and relabeling the nodes of `other` that were not merged.
+
+        Parameters
+        ----------
+        other : OpenGraph
+            Open graph to be composed with `self`.
+        mapping: dict[int, int]
+            Partial relabelling of the nodes in `other`, with `keys` and `values` denoting the old and new node labels, respectively.
+
+        Returns
+        -------
+        og: OpenGraph
+            composed open graph
+        mapping_complete: dict[int, int]
+            Complete relabelling of the nodes in `other`, with `keys` and `values` denoting the old and new node label, respectively.
+
+        Notes
+        -----
+        Let's denote :math:`\{G(V_1, E_1), I_1, O_1\}` the open graph `self`, :math:`\{G(V_2, E_2), I_2, O_2\}` the open graph `other`, :math:`\{G(V, E), I, O\}` the resulting open graph `og` and `{v:u}` an element of `mapping`.
+
+        We define :math:`V, U` the set of nodes in `mapping.keys()` and `mapping.values()`, and :math:`M = U \cap V_1` the set of merged nodes.
+
+        The open graph composition requires that
+        - :math:`V \subseteq V_2`.
+        - If both `v` and `u` are measured, the corresponding measurements must have the same plane and angle.
+         The returned open graph follows this convention:
+        - :math:`I = (I_1 \cup I_2) \setminus M \cup (I_1 \cap I_2 \cap M)`,
+        - :math:`O = (O_1 \cup O_2) \setminus M \cup (O_1 \cap O_2 \cap M)`,
+        - If only one node of the pair `{v:u}` is measured, this measure is assigned to :math:`u \in V` in the resulting open graph.
+        - Input (and, respectively, output) nodes in the returned open graph have the order of the open graph `self` followed by those of the open graph `other`. Merged nodes are removed, except when they are input (or output) nodes in both open graphs, in which case, they appear in the order they originally had in the graph `self`.
+        """
+        if not mapping.keys() <= other.inside.nodes:
```

Comment:
Let's add surrounding brackets for readability.

```suggestion
        if not (mapping.keys() <= other.inside.nodes):
```